### PR TITLE
8324723: GHA: Upgrade some actions to avoid deprecated Node 16

### PR DESCRIPTION
**Pull request by gdams**
This backport is written from scratch, as GHA workflows in jdk8u differ significantly from newer jdk versions.

This will remove all warnings about deprecated Node.js versions

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] [JDK-8324723](https://bugs.openjdk.org/browse/JDK-8324723) needs maintainer approval
- [x] [JDK-8315863](https://bugs.openjdk.org/browse/JDK-8315863) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8324723](https://bugs.openjdk.org/browse/JDK-8324723): GHA: Upgrade some actions to avoid deprecated Node 16 (**Enhancement** - P4 - Approved)
 * [JDK-8315863](https://bugs.openjdk.org/browse/JDK-8315863): [GHA] Update checkout action to use v4 (**Enhancement** - P4 - Approved)

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Zdenek Zambersky](https://openjdk.org/census#zzambers) (@zzambers - Committer) 🔄 Re-review required (review applies to [ec96589f](https://git.openjdk.org/jdk8u-dev/pull/546/files/ec96589f7246d5ca7a48d3b351791a578ebb1db0))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u-dev.git pull/546/head:pull/546` \
`$ git checkout pull/546`

Update a local copy of the PR: \
`$ git checkout pull/546` \
`$ git pull https://git.openjdk.org/jdk8u-dev.git pull/546/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 546`

View PR using the GUI difftool: \
`$ git pr show -t 546`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u-dev/pull/546.diff">https://git.openjdk.org/jdk8u-dev/pull/546.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk8u-dev/pull/546#issuecomment-2225328149)

### CODE REVIEW

**Review by jerboaa — APPROVED** [review]
LGTM. Thanks for fixing.

**Comment by zzambers** [comment]
I remember I was looking a bit into this in the past. And there were some problems with `actions/upload-artifact@v4`, as there seem to be some [braking change](https://github.com/actions/upload-artifact/issues/478) in v4 (from v3).

See [errors](https://github.com/gdams/jdk8u-dev/actions/runs/9906731382) in sumary (annotations):
```
Failed to CreateArtifact: Received non-retryable error: Failed request: (409) Conflict: an artifact with this name already exists on the workflow run
```

It seems to have effect on produced artifacts (bottom of the page), compare with e.g.:
https://github.com/zzambers/jdk8u-dev/actions/runs/9858657012

With existing actions it produces just one `test-results` artifact, which is relatively small (<200MiB), while with updated 
actions there are some additional (leftover?) artifacts and test-results is huge (1.49GiB).

Looking at contents of `test-results` with updated actions, it's is wrong. It does not contain results of all tests, as it should, and wrongly contains built jdk binaries...

However I haven't looked into detail, why that is. Or how difficult would it be to fix this.

**Comment by zzambers** [comment]
It seems there were some [significant changes](https://github.com/actions/upload-artifact?tab=readme-ov-file#v4---whats-new) in artifact actions v4. (causing compatibility problems with previous versions)

I took a deeper look and there are actually 2 issues here:

**1. Upload of test results**
 -  artifact name is not specified (in upload-artifact) for test results uploads -> default artifact name ("artifact") is then used
 - upload-artifact v4 can no longer upload multiple times (append) to the same artifact -> error messages, missing archives with results
 - can be easily [fixed](https://github.com/zzambers/jdk8u-dev/commit/b5abfa7111d1c1360de8b43e5d7a0200c77ef569) by explicitly specifying artifact name for testresults
 
**2. Artifact removal does not work**
  - [current code](https://github.com/openjdk/jdk8u-dev/blob/f231e27d90dd282355b5776653af20ca04f75591/.github/workflows/submit.yml#L1589) to remove unnecessary artifacts no longer works with artifacts uploaded by `upload-artifact@v4` 
  - also affects newer jdks, I have crated bug and fix (see [JDK-8336928](https://bugs.openjdk.org/browse/JDK-8336928))

**Comment by gdams** [comment]
> * JDK-8336928

Sorry for the delay, I've applied the fixes @zzambers

**Comment by zzambers** [comment]
> > * JDK-8336928
> 
> Sorry for the delay, I've applied the fixes @zzambers

Thanks, looks good. (no more problems with uploading archives with test results)

There should then be followup backport of [JDK-8336928](https://bugs.openjdk.org/browse/JDK-8336928) to also fix problem with removing unnecessary artifacts (uploaded by newer actions), but unfortunately, I am still waiting for review for that fix. :(

**Review by zzambers — APPROVED** [review]

**Comment by gdams** [comment]
@jerboaa I need to request another review please

File: .github/workflows/submit.yml
Comment (gdams) [inline .github/workflows/submit.yml:1624]:
```suggestion
        uses: actions/upload-artifact@v4
```

**Review by jerboaa — APPROVED** [review]
Seems fine. I'll rely on @zzambers for checking this doesn't break any other things.

**Comment by gdams** [comment]
/approval request backport to dodge GHA dependencies deprecations. GHA works without problems in target repo.

**Comment by gdams** [comment]
> @gdams Your change (at version [efe7e7a](https://github.com/openjdk/jdk8u-dev/commit/efe7e7a955a15ae168af67e759c0c1926dfa1d1f)) is now ready to be sponsored by a Committer.

I am a committer 🤷🏼 

/sponsor

**Comment by gdams** [comment]
> @gdams Only [Committers](https://openjdk.org/bylaws#committer) are allowed to sponsor changes.

bot seems broken, @jerboaa would you mind doing the honours?

**Comment by jerboaa** [comment]
> > @gdams Only [Committers](https://openjdk.org/bylaws#committer) are allowed to sponsor changes.
> 
> bot seems broken, @jerboaa would you mind doing the honours?

Looking at the [census](https://openjdk.org/census#gdams) you'll find that you are and jdk-updates project committer. This is project [JDK 8 Update Releases](https://openjdk.org/projects/jdk8u/). So the bot has it correct. I'll sponsor it.

**Comment by jerboaa** [comment]
GHA failures are unrelated to this patch.

/sponsor

**Pull request closed without merging**
